cmd/ipfs/idxprovider.go

```diff
@@ -0,0 +1,442 @@
+			return err
+		}
+	}
+	return err
+}
+
+func registerWithIndexers(cctx idxProviderContext, node idxProviderNode, indexerNode string) error {
+	providerlog.Debugf("registerWithIndexers")
+
+	client, err := httpc.New(indexerNode)
+	if err != nil {
+		return err
+	}
+
+	peerID := node.PeerHost().ID()
+	privKey := node.PeerHost().Peerstore().PrivKey(node.PeerHost().ID())
+	providerlog.Debugf("calling storetheindex register")
+	return client.Register(cctx.Context(), peerID, privKey, node.Addrs())
+}
+
+func advertiseAllCids(ctx context.Context, node idxProviderNode) error {
+	providerlog.Debugf("Advertising latest CIDs")
+
+	allCids, err := node.BlockStore().AllKeysChan(ctx)
+	if err != nil {
+		return err
+	}
+
+	var cnt = 0
+	var cids []cid.Cid
+
+	for c := range allCids {
+		cids = append(cids, c)
+		cids[cnt] = c
+		cnt++
+		if cnt >= maxCidsPerAdv {
+			err = doAdvertisement(ctx, node, cids)
+			if err != nil {
+				providerlog.Errorf("Error advertising: %s", err)
+				return err
+			}
+			// reset array for next window
+			cids = cids[:0]
+			cnt = 0
+		}
+
+	}
+	// advertise the leftovers
+	err = doAdvertisement(ctx, node, cids)
+	if err != nil {
+		providerlog.Errorf("Error advertising: %s", err)
+	}
+	return err
+}
+
+func doAdvertisement(ctx context.Context, node idxProviderNode, cids []cid.Cid) error {
+	providerlog.Debugf("generating chunks for %v cids", len(cids))
+	lnk, contextID, err := generateChunks(ctx, cids)
+	if err != nil {
+		providerlog.Debug("Error generating chunks")
+		return err
+	}
+	if lnk == nil {
+		providerlog.Debug("Nothing to advertise this time")
+		return nil
+	}
+
+	cidsLnk := lnk.(cidlink.Link)
+
+	providerlog.Debugf("generating bogus metadata")
+	// 0x0900 is the bitswap protocol ID
+	metadata := stiapi.Metadata{
+		ProtocolID: 0x0900,
+	}
+
+	peerID := node.PeerHost().ID()
+	privKey := node.PeerHost().Peerstore().PrivKey(node.PeerHost().ID())
+
+	// Check for cid.Undef for the previous link. If this is the case, then
+	// this means there is a "cid too short" error in IPLD links serialization.
+	previousLnk, err := getPreviousAdvertisementLink(ctx)
+	if err != nil {
+		providerlog.Warnf("Error getting previous adv link, sending nil:  %s", err)
+		previousLnk = nil
+	}
+
+	providerlog.Debugf("creating new adv")
+	adv, err := schema.NewAdvertisement(privKey, previousLnk, cidsLnk,
+		contextID, metadata, false, peerID.String(), node.Addrs())
+	if err != nil {
+		providerlog.Debugf("got error from new advertisement: %v", err)
+		return err
+	}
+
+	providerlog.Debugf("creating new adv link")
+	adLnk, err := schema.AdvertisementLink(lsys, adv)
+	if err != nil {
+		providerlog.Errorf("Error generating advertisement link: %s", err)
+		return err
+	}
+
+	c := adLnk.ToCid()
+
+	// Store latest advertisement published from the chain
+	providerlog.Infow("Storing advertisement locally", "cid", c.String())
+	err = dataStore.Put(ctx, datastore.NewKey(latestAdvKey), c.Bytes())
```

Comment:
Note: If you wanted to keep your advertisements safe from GC it seems pretty straightforward to do. You just have to pass the root CID (as found from the `latestAdvKey`) into the list of roots that GC should traverse and mark as safe from GC because IIUC the advertisements are all linked together in a single DAG.